### PR TITLE
Anvil options for repairing and damaging

### PR DESCRIPTION
**Pull request by 12emin34**
This patch allows for repairing anvils using iron ingots and damaging them using obsidian.

### CODE REVIEW

File: patches/server/0270-Anvil-repair-damage-options.patch [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+         } else {
+diff --git a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
+index 7c131da608b91555cd5293da5b90de0ee9c10f9e..101e949555b309f10e466ffe475bf40db565465b 100644
+--- a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
++++ b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
+@@ -699,8 +699,12 @@ public class PurpurWorldConfig {
+     }
+ 
+     public boolean anvilAllowColors = false;
++    public boolean anvilCanRepairUsingIngots = false;
```
Comment (BillyGalbreath) [inline patches/server/0270-Anvil-repair-damage-options.patch]: what if we made this an int so we can specify how many ingots do a repair? a single ingot seems really cheap to me. a value of 0 or less would consider the feature disabled.

File: patches/server/0270-Anvil-repair-damage-options.patch [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
++import net.minecraft.world.item.Items;
+ import net.minecraft.world.item.context.BlockPlaceContext;
+ import net.minecraft.world.level.BlockGetter;
+ import net.minecraft.world.level.Level;
+@@ -55,6 +56,33 @@ public class AnvilBlock extends FallingBlock {
+ 
+     @Override
+     public InteractionResult use(BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
++        // Purpur start - repairable/damageable anvils
++        if (player.getItemInHand(hand).is(Items.IRON_INGOT) && world.purpurConfig.anvilCanRepairUsingIngots) {
```
Comment (BillyGalbreath) [inline patches/server/0270-Anvil-repair-damage-options.patch]: flip these checks so we're not doing 2 checks on servers where this is disabled.

File: patches/server/0270-Anvil-repair-damage-options.patch [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
++            if (state.is(Blocks.DAMAGED_ANVIL)) {
++                world.setBlock(pos, Blocks.CHIPPED_ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
++                if (!player.getAbilities().instabuild) player.getItemInHand(hand).shrink(1);
++                return InteractionResult.CONSUME;
++            } else if (state.is(Blocks.CHIPPED_ANVIL)) {
++                world.setBlock(pos, Blocks.ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
++                if (!player.getAbilities().instabuild) player.getItemInHand(hand).shrink(1);
++                return InteractionResult.CONSUME;
++            }
++        } else if (player.getItemInHand(hand).is(Items.OBSIDIAN) && world.purpurConfig.anvilCanDamageUsingObsidian) {
```
Comment (BillyGalbreath) [inline patches/server/0270-Anvil-repair-damage-options.patch]: flip these checks so we're not doing 2 checks on servers where this is disabled.

File: patches/server/0270-Anvil-repair-damage-options.patch [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
++                return InteractionResult.CONSUME;
++            } else if (state.is(Blocks.CHIPPED_ANVIL)) {
++                world.setBlock(pos, Blocks.DAMAGED_ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
++                if (!player.getAbilities().instabuild) player.getItemInHand(hand).shrink(1);
++                return InteractionResult.CONSUME;
++            } else if (state.is(Blocks.ANVIL)) {
++                world.setBlock(pos, Blocks.CHIPPED_ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
++                if (!player.getAbilities().instabuild) player.getItemInHand(hand).shrink(1);
++                return InteractionResult.CONSUME;
++            }
```
Comment (BillyGalbreath) [inline patches/server/0270-Anvil-repair-damage-options.patch]: there's a lot of repeat code in here

```suggestion
+            if (state.is(Blocks.DAMAGED_ANVIL)) {
+                world.destroyBlock(pos, false);
+            } else if (state.is(Blocks.CHIPPED_ANVIL)) {
+                world.setBlock(pos, Blocks.DAMAGED_ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
+            } else if (state.is(Blocks.ANVIL)) {
+                world.setBlock(pos, Blocks.CHIPPED_ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
+            }
+            if (!player.getAbilities().instabuild) player.getItemInHand(hand).shrink(1);
+            return InteractionResult.CONSUME;
```

File: patches/server/0270-Anvil-repair-damage-options.patch [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
++        if (player.getItemInHand(hand).is(Items.IRON_INGOT) && world.purpurConfig.anvilCanRepairUsingIngots) {
++            if (state.is(Blocks.DAMAGED_ANVIL)) {
++                world.setBlock(pos, Blocks.CHIPPED_ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
++                if (!player.getAbilities().instabuild) player.getItemInHand(hand).shrink(1);
++                return InteractionResult.CONSUME;
++            } else if (state.is(Blocks.CHIPPED_ANVIL)) {
++                world.setBlock(pos, Blocks.ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
++                if (!player.getAbilities().instabuild) player.getItemInHand(hand).shrink(1);
++                return InteractionResult.CONSUME;
++            }
```
Comment (BillyGalbreath) [inline patches/server/0270-Anvil-repair-damage-options.patch]: there's a lot of repeat code in here

```suggestion
+            if (state.is(Blocks.DAMAGED_ANVIL)) {
+                world.setBlock(pos, Blocks.CHIPPED_ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
+            } else if (state.is(Blocks.CHIPPED_ANVIL)) {
+                world.setBlock(pos, Blocks.ANVIL.defaultBlockState().setValue(FACING, state.getValue(FACING)), 2);
+            } else if (state.is(Blocks.CHIPPED_ANVIL)) {
+                return InteractionResult.CONSUME
+            }
+            if (!player.getAbilities().instabuild) player.getItemInHand(hand).shrink(1);
+            return InteractionResult.CONSUME;
```

File: patches/server/0270-Anvil-repair-damage-options.patch [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+diff --git a/src/main/java/net/minecraft/world/level/block/AnvilBlock.java b/src/main/java/net/minecraft/world/level/block/AnvilBlock.java
+index 61cadc946954a1e92bd9b395b87dcb83eef786c4..aaa4aa929140a7e0630360caa3195a9264006e16 100644
+--- a/src/main/java/net/minecraft/world/level/block/AnvilBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/AnvilBlock.java
+@@ -15,6 +15,7 @@ import net.minecraft.world.entity.item.FallingBlockEntity;
+ import net.minecraft.world.entity.player.Player;
+ import net.minecraft.world.inventory.AnvilMenu;
+ import net.minecraft.world.inventory.ContainerLevelAccess;
++import net.minecraft.world.item.Items;
```
Comment (BillyGalbreath) [inline patches/server/0270-Anvil-repair-damage-options.patch]: you can inline this. its only used twice.

**Review by BillyGalbreath — CHANGES_REQUESTED** [review]

**Comment by 12emin34** [comment]
I have made the requested changes and added sounds since it was kinda dull without them.

**Review by BillyGalbreath — APPROVED** [review]

**Pull request merged**